File: 1615.maximal-network-rank.py
#
# @lc app=leetcode id=1615 lang=python3
#
# [1615] Maximal Network Rank
#

# @lc code=start
from typing import List


class Solution:
    '''
    Compute degree of each node
    Then for each pair, computer deg1 + deg2  (- 1 if they are linked)    
    '''
    def maximalNetworkRank(self, n: int, roads: List[List[int]]) -> int:
        degrees = [0]*n
        for node1, node2 in roads:
            degrees[node1] += 1
            degrees[node2] += 1
        result, roads_set = 0, set()
        for road in roads:
            roads_set.add(tuple(road))
        for node1 in range(n-1):
            for node2 in range(node1 + 1, n):
                if (node1, node2) in roads_set or (node2, node1) in roads_set:
                    result = max(result, degrees[node1] + degrees[node2] - 1)
                else:
                    result = max(result, degrees[node1] + degrees[node2])
        return result
# @lc code=end

File: 1615.maximal-network-rank.cpp
/*
 * @lc app=leetcode id=1615 lang=cpp
 *
 * [1615] Maximal Network Rank
 */

// @lc code=start
#include <vector>
#include <unordered_set>
#include <algorithm>
#include <tuple>

class Solution {
public:
    int maximalNetworkRank(int n, std::vector<std::vector<int>>& roads) {
        std::vector<int> degrees(n, 0);
        std::vector<std::vector<bool>> connected(n, std::vector<bool>(n, false));
        for (const auto& road : roads) {
            degrees[road[0]]++;
            degrees[road[1]]++;
            connected[std::min(road[0], road[1])][std::max(road[0], road[1])] = true;
        }

        int result = 0;
        for (int node1 = 0; node1 < n - 1; node1++) {
            for (int node2 = node1 + 1; node2 < n; node2++) {
                if (connected[node1][node2]) {
                    result = std::max(result, degrees[node1] + degrees[node2] - 1);
                } else {
                    result = std::max(result, degrees[node1] + degrees[node2]);
                }
            }
        }

        return result;
    }
};
// @lc code=end

